File: cliente.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <winsock2.h>
#include <windows.h>
#include "include/cJSON.h"

#pragma comment(lib, "ws2_32.lib")

#define BUFFER_SIZE 1024

int get_num_processors() {
    SYSTEM_INFO sysinfo;
    GetSystemInfo(&sysinfo);
    return sysinfo.dwNumberOfProcessors;
}

long get_free_ram() {
    MEMORYSTATUSEX memInfo;
    memInfo.dwLength = sizeof(MEMORYSTATUSEX);
    GlobalMemoryStatusEx(&memInfo);
    return memInfo.ullAvailPhys / (1024 * 1024); // MB
}

long get_free_disk_space() {
    ULARGE_INTEGER freeBytesAvailable;
    GetDiskFreeSpaceEx("C:\\", &freeBytesAvailable, NULL, NULL);
    return freeBytesAvailable.QuadPart / (1024 * 1024 * 1024); // GB
}

float get_cpu_temperature() {
    FILE *fp;
    char path[1035];
    float temperatura = -273.15; // Valor padrão caso falhe

    fp = _popen("powershell -Command \"(Get-WmiObject MSAcpi_ThermalZoneTemperature -Namespace 'root/wmi').CurrentTemperature\"", "r");
    if (fp) {
        if (fgets(path, sizeof(path), fp) != NULL) {
            temperatura = atof(path) / 10.0 - 273.15; // Celsius
        }
        _pclose(fp);
    }

    return temperatura;
}

float calcular_media(int cpu, long ram, long disk, float temperatura) {
    return (cpu + ram + disk + temperatura) / 4.0;
}

int main() {
    WSADATA wsaData;
    SOCKET clientSocket;
    struct sockaddr_in serverAddr;
    char buffer[BUFFER_SIZE];

    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        printf("Erro ao inicializar Winsock.\n");
        return 1;
    }

    clientSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (clientSocket == INVALID_SOCKET) {
        printf("Erro ao criar socket.\n");
        WSACleanup();
        return 1;
    }

    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(8080);
    serverAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    if (connect(clientSocket, (struct sockaddr *)&serverAddr, sizeof(serverAddr)) < 0) {
        printf("Erro ao conectar ao servidor.\n");
        closesocket(clientSocket);
        WSACleanup();
        return 1;
    }

    // Coletar informações do sistema
    int cpu = get_num_processors();
    long ram = get_free_ram();
    long disk = get_free_disk_space();
    float temp = get_cpu_temperature();
    float media = calcular_media(cpu, ram, disk, temp);

    // Criar JSON com as informações
    cJSON *json = cJSON_CreateObject();
    cJSON_AddNumberToObject(json, "cpu", cpu);
    cJSON_AddNumberToObject(json, "ram", ram);
    cJSON_AddNumberToObject(json, "disk", disk);
    cJSON_AddNumberToObject(json, "temperature", temp);
    cJSON_AddNumberToObject(json, "average", media);

    char *jsonString = cJSON_Print(json);
    send(clientSocket, jsonString, strlen(jsonString), 0);

    cJSON_Delete(json);
    free(jsonString);

    // Menu de interação
    while (1) {
        int bytesReceived = recv(clientSocket, buffer, BUFFER_SIZE, 0);
        if (bytesReceived <= 0) {
            printf("Servidor desconectou.\n");
            break;
        }

        buffer[bytesReceived] = '\0';
        printf("%s", buffer);

        printf("Escolha uma opção: ");
        fgets(buffer, BUFFER_SIZE, stdin);
        send(clientSocket, buffer, strlen(buffer), 0);

        if (strncmp(buffer, "sair", 4) == 0) break;
    }

    closesocket(clientSocket);
    WSACleanup();
    return 0;
}

File: cliente.py
import socket
import json
import psutil

class ComputerInfo:
    """Classe responsável por coletar e armazenar informações do sistema."""

    def __init__(self):
        self.cpu = self.get_num_processors()
        self.ram = self.get_free_ram()
        self.disk = self.get_free_disk_space()
        self.temp = self.get_cpu_temperature()
        self.media = self.calcular_media()

    def get_num_processors(self):
        """Retorna o número de processadores físicos."""
        return psutil.cpu_count(logical=False)

    def get_free_ram(self):
        """Retorna a quantidade de RAM livre em MB."""
        return psutil.virtual_memory().available // (1024 * 1024)

    def get_free_disk_space(self):
        """Retorna o espaço livre em disco em GB."""
        disk = psutil.disk_usage('/')
        return disk.free // (1024 * 1024 * 1024)

    def get_cpu_temperature(self):
        """Retorna a temperatura da CPU em Celsius (se disponível)."""
        try:
            import wmi
            w = wmi.WMI(namespace="root\\wmi")
            temperatures = w.MSAcpi_ThermalZoneTemperature()
            if temperatures:
                return (temperatures[0].CurrentTemperature / 10.0) - 273.15
        except Exception as e:
            print(f"Erro ao obter temperatura: {e}")
        return None 

    def calcular_media(self):
        """Calcula a média simples dos dados."""
        valores = [self.cpu, self.ram, self.disk]
        if self.temp is not None:
            valores.append(self.temp)  
        return sum(valores) / len(valores)

    def formatar_dados(self):
        """Formata os dados do sistema com unidades de medida."""
        return {
            "Processadores": f"{self.cpu} und",
            "RAM Disponível": f"{self.ram} MB",
            "Espaço em Disco Livre": f"{self.disk} GB",
            "Temperatura": f"{self.temp:.2f} °C" if self.temp is not None else "N/A",
            "Média dos Dados": f"{self.media:.2f}"
        }

class Client:
    """Classe responsável pela comunicação com o servidor."""

    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.system_info = ComputerInfo()
        self.key = "minhachave"  

    def xor_cipher(self, input_data):
        """Aplica a criptografia XOR aos dados."""
        output_data = bytearray()
        key_len = len(self.key)
        for i in range(len(input_data)):
            output_data.append(input_data[i] ^ ord(self.key[i % key_len]))
        return bytes(output_data)

    def connect(self):
        """Estabelece a conexão com o servidor e envia os dados."""
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
            try:
                client_socket.connect((self.host, self.port))
                print(f"🔗 Conectado ao servidor {self.host}:{self.port}")

               
                json_data = json.dumps(self.system_info.formatar_dados()).encode()

                # Criptografa os dados 
                encrypted_data = self.xor_cipher(json_data)
                print(f"🔒 Dados criptografados: {encrypted_data.hex()}")  
                client_socket.send(encrypted_data)
                print("✅ Dados do sistema enviados ao servidor.")

                # Menu de interação
                while True:
                    response = client_socket.recv(1024).decode()
                    if not response:
                        print("⚠️ Servidor desconectou.")
                        break

                    print(response)

                    
                    choice = input("Escolha uma opção: ")
                    client_socket.send(choice.encode())

                    if choice.strip().lower() == "sair":
                        break

            except Exception as e:
                print(f"❌ Erro ao conectar ou comunicar com o servidor: {e}")

if __name__ == "__main__":
    client = Client("192.168.15.14", 8080)
    client.connect()
